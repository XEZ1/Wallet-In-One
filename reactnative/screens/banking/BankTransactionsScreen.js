import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View, ScrollView, Button, TextInput, Alert, FlatList , TouchableOpacity, Image} from 'react-native';
import { useState, useEffect } from 'react';
import { auth_get} from '../../authentication'
import { useIsFocused } from '@react-navigation/native';
import Loading from './Loading'
import { useTheme } from 'reactnative/src/theme/ThemeProvider'

function BankTransactionsScreen({ route, navigation }) {
  const [ isLoading, setIsLoading ] = useState(true)
  const [ bankData, setBankData ] = useState([])
  const isFocused = useIsFocused();
  const {dark, colors, setScheme} = useTheme();

  useEffect(() =>{
    const fetchData = async () => {
        console.log('fetch bank data')
        if (route.params == undefined){
          response = await auth_get('/banking/transactions/')
        }
        else{
          response = await auth_get('/banking/transactions/' + route.params.accountID +'/')
        }
        
        if (response.status == 200){
            setIsLoading(false)
            response.body.sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp))
            setBankData(response.body)
        }
    }
    if (isFocused){fetchData()}
  }, [isFocused])

  const displayDate = (timestamp) => {
    const date = new Date(timestamp);
    return  date.toLocaleDateString();
  };

  const displayTime = (timestamp) => {
    const date = new Date(timestamp);
    return  date.toLocaleTimeString();
  };

  const styles = StyleSheet.create({
    container: {
      width: '100%',
      paddingLeft: 20,
      paddingRight: 20,
      paddingBottom: 0,
      borderWidth: 1,
      borderRadius: 5,
      borderColor: dark ? colors.background : '#ddd',
      overflow: 'hidden',
    },
    row:{
      flexDirection: 'row',
      alignItems: 'center',
    },
    olditem:{
      padding: 20,
    },
    text:{
      color: colors.text
    },
    item: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      paddingVertical: 8,
      borderBottomWidth: 1,
      borderBottomColor: 'lightgray',
    },
    leftContainer: {
      flex: 1,
    },
    rightContainer: {
      marginLeft: 10,
      alignItems: 'flex-end',
    },
    name: {
      fontWeight: 'bold',
      fontSize: 14,
      color: colors.text,
    },
    date: {
      fontSize: 14,
      color: 'gray',
    },
    amount: {
      fontSize: 16,
      fontWeight: 'bold',
    },
    time: {
      fontSize: 14,
      color: 'gray',
    },
  });

  const TransactionItem = ({ item, name, date, amount, time }) => {
    const amountColor = amount >= 0 ? 'green' : 'red';
    return (
      <View style={styles.item}>
        <View style={styles.leftContainer}>
          <ScrollView horizontal={true}>
            <Text style={styles.name}>{name}</Text>
          </ScrollView>
          <Text style={styles.date}>{date} at {time}</Text>
        </View>
        <View style={styles.rightContainer}>
          <Text style={[styles.amount, { color: amountColor }]}>{item.formatted_amount.string}</Text>
        </View>
      </View>
    );
  };

  if (isLoading){
    return <Loading/>
}

  return (
    <View style={{flex:1,  margin: 4, marginBottom: 54}} >
              <View style={[styles.container, {backgroundColor: colors.background}]}>
                  <FlatList data={bankData} renderItem={({item, index}) =>{
                      return (
                        <TransactionItem
                          item={item}
                          name={item.info}
                          amount={item.amount}
                          date={displayDate(item.time)}
                          time={displayTime(item.time)} />)
                      }}
                      ListEmptyComponent={<Text style={styles.text}>{'\nYou have no bank accounts\n'}</Text>}
                  />
              </View>
      </View>
  );
}


export default BankTransactionsScreen