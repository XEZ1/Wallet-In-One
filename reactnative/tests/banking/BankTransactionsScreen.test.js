import React from 'react';
import { render, screen, fireEvent, act, within, waitFor} from '@testing-library/react-native';
import BankTransactionsScreen from '../../screens/banking/BankTransactionsScreen'

global.fetch =  jest.fn( async (api, data ) => {

    const transactions = [{
        "id": "abc",
        "time": "2023-03-12T11:42:32Z",
        "amount_currency": "GBP",
        "amount": "-10.00",
        "info": "t1",
        "account": "def",
        "formatted_amount": {
            "string": "-£10.00",
            "currency": "GBP",
            "amount": "-10.00"
        }
    },
    {
        "id": "def",
        "time": "2023-03-11T11:46:36Z",
        "amount_currency": "GBP",
        "amount": "5.00",
        "info": "t2",
        "account": "def",
        "formatted_amount": {
            "string": "£5.00",
            "currency": "GBP",
            "amount": "5.00"
        }
    }]

    function hasAccountID(url) {
        const pattern = /\/transactions\/[^/]+(\/|$)/;
        return pattern.test(url);
    }
    if (hasAccountID(api)){
        return Promise.resolve({
            status: 200, json: () => [transactions[1]]
        })
    }
    else{
        return Promise.resolve({
            status: 200, json: () => transactions
        })
    }

})

describe('<BankTransactionsScreen />', () => {
    it('test transactions with multiple accounts', async () => {
        

        const snapshot = render(<BankTransactionsScreen route={{params: undefined}}/>);

        await act( async () => {
            
            await waitFor( () => {
                const activityIndicator = screen.UNSAFE_queryByType('ActivityIndicator');
                expect(activityIndicator).toBeNull();
            })

        });

        expect(screen.queryAllByText('t1')).toHaveLength(1)
        expect(screen.queryAllByText('-£10.00')).toHaveLength(1)
        expect(screen.queryAllByText('t2')).toHaveLength(1)
        expect(screen.queryAllByText('£5.00')).toHaveLength(1)


    }
    )

    it('test transactions with 1 account', async () => {

        const snapshot = render(<BankTransactionsScreen route={{params: 'abc'}}/>);

        await act( async () => {
            
            await waitFor( () => {
                const activityIndicator = screen.UNSAFE_queryByType('ActivityIndicator');
                expect(activityIndicator).toBeNull();
            })

        });

        expect(screen.queryAllByText('t1')).toHaveLength(0)
        expect(screen.queryAllByText('-£10.00')).toHaveLength(0)
        expect(screen.queryAllByText('t2')).toHaveLength(1)
        expect(screen.queryAllByText('£5.00')).toHaveLength(1)

    }
    )
});