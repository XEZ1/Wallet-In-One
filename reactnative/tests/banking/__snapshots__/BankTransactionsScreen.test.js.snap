// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BankTransactionsScreen /> snapshot test 1`] = `
<View
  style={
    {
      "flex": 1,
      "margin": 4,
      "marginBottom": 54,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "white",
          "borderColor": "#ddd",
          "borderRadius": 5,
          "borderWidth": 1,
          "overflow": "hidden",
          "paddingBottom": 0,
          "width": "100%",
        },
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <View
          style={
            {
              "paddingLeft": 20,
              "paddingRight": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
              }
            }
          >
            
You have no bank accounts

          </Text>
        </View>
      }
      data={
        [
          {
            "data": [
              {
                "account": "def",
                "amount": "-10.00",
                "amount_currency": "GBP",
                "formatted_amount": {
                  "amount": "-10.00",
                  "currency": "GBP",
                  "string": "-£10.00",
                },
                "id": "abc",
                "info": "t1",
                "time": "2023-03-12T11:42:32Z",
              },
              {
                "account": "def",
                "amount": "-10.00",
                "amount_currency": "GBP",
                "formatted_amount": {
                  "amount": "10.00",
                  "currency": "GBP",
                  "string": "£10.00",
                },
                "id": "def",
                "info": "t2",
                "time": "2023-03-11T11:46:36Z",
              },
            ],
            "title": "March 2023",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={
        [
          0,
        ]
      }
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <Text
            style={
              {
                "backgroundColor": "#f5f5f5",
                "color": "black",
                "fontWeight": "bold",
                "paddingBottom": 6,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 6,
              }
            }
          >
            March 2023
          </Text>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingVertical": 8,
                  },
                  {
                    "borderBottomColor": "lightgray",
                    "borderBottomWidth": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                  }
                }
              >
                <RCTScrollView
                  horizontal={true}
                >
                  <View>
                    <Text
                      style={
                        {
                          "color": "black",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      t1
                    </Text>
                  </View>
                </RCTScrollView>
                <Text
                  style={
                    {
                      "color": "gray",
                      "fontSize": 14,
                    }
                  }
                >
                  12/03/2023
                   at 
                  11:42:32
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-end",
                    "marginLeft": 10,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "fontWeight": "bold",
                      },
                      {
                        "color": "red",
                      },
                    ]
                  }
                >
                  -£10.00
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingVertical": 8,
                  },
                  {},
                ]
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                  }
                }
              >
                <RCTScrollView
                  horizontal={true}
                >
                  <View>
                    <Text
                      style={
                        {
                          "color": "black",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      t2
                    </Text>
                  </View>
                </RCTScrollView>
                <Text
                  style={
                    {
                      "color": "gray",
                      "fontSize": 14,
                    }
                  }
                >
                  11/03/2023
                   at 
                  11:46:36
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-end",
                    "marginLeft": 10,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "fontWeight": "bold",
                      },
                      {
                        "color": "red",
                      },
                    ]
                  }
                >
                  £10.00
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        />
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`<BankTransactionsScreen /> snapshot test with account id 1`] = `
<View
  style={
    {
      "flex": 1,
      "margin": 4,
      "marginBottom": 54,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "white",
          "borderColor": "#ddd",
          "borderRadius": 5,
          "borderWidth": 1,
          "overflow": "hidden",
          "paddingBottom": 0,
          "width": "100%",
        },
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <View
          style={
            {
              "paddingLeft": 20,
              "paddingRight": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
              }
            }
          >
            
You have no bank accounts

          </Text>
        </View>
      }
      data={
        [
          {
            "data": [
              {
                "account": "def",
                "amount": "-10.00",
                "amount_currency": "GBP",
                "formatted_amount": {
                  "amount": "-10.00",
                  "currency": "GBP",
                  "string": "-£10.00",
                },
                "id": "abc",
                "info": "t1",
                "time": "2023-03-12T11:42:32Z",
              },
              {
                "account": "def",
                "amount": "-10.00",
                "amount_currency": "GBP",
                "formatted_amount": {
                  "amount": "10.00",
                  "currency": "GBP",
                  "string": "£10.00",
                },
                "id": "def",
                "info": "t2",
                "time": "2023-03-11T11:46:36Z",
              },
            ],
            "title": "March 2023",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={
        [
          0,
        ]
      }
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <Text
            style={
              {
                "backgroundColor": "#f5f5f5",
                "color": "black",
                "fontWeight": "bold",
                "paddingBottom": 6,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 6,
              }
            }
          >
            March 2023
          </Text>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingVertical": 8,
                  },
                  {
                    "borderBottomColor": "lightgray",
                    "borderBottomWidth": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                  }
                }
              >
                <RCTScrollView
                  horizontal={true}
                >
                  <View>
                    <Text
                      style={
                        {
                          "color": "black",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      t1
                    </Text>
                  </View>
                </RCTScrollView>
                <Text
                  style={
                    {
                      "color": "gray",
                      "fontSize": 14,
                    }
                  }
                >
                  12/03/2023
                   at 
                  11:42:32
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-end",
                    "marginLeft": 10,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "fontWeight": "bold",
                      },
                      {
                        "color": "red",
                      },
                    ]
                  }
                >
                  -£10.00
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingVertical": 8,
                  },
                  {},
                ]
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                  }
                }
              >
                <RCTScrollView
                  horizontal={true}
                >
                  <View>
                    <Text
                      style={
                        {
                          "color": "black",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      t2
                    </Text>
                  </View>
                </RCTScrollView>
                <Text
                  style={
                    {
                      "color": "gray",
                      "fontSize": 14,
                    }
                  }
                >
                  11/03/2023
                   at 
                  11:46:36
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-end",
                    "marginLeft": 10,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "fontWeight": "bold",
                      },
                      {
                        "color": "red",
                      },
                    ]
                  }
                >
                  £10.00
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        />
      </View>
    </RCTScrollView>
  </View>
</View>
`;
